File: javascript/util/permutation.js
const digits = require(`./digits`);

function isPermutation(num, testNum) {
  const numDigits = digits.getDigitsReversed(num);
  const testNumDigits = digits.getDigitsReversed(testNum);

  if (numDigits.length !== testNumDigits.length) {
    return false;
  }

  // create hashmap of digit counts in first array
  const digitCounts = new Map();
  for (let i = 0; i < numDigits.length; i++) {
    const digit = numDigits[i];
    if (!digitCounts.has(digit)) {
      digitCounts.set(digit, 1);
    } else {
      digitCounts.set(digit, digitCounts.get(digit) + 1);
    }
  }

  // check digit counts in the second array to make sure they match
  for (let i = 0; i < testNumDigits.length; i++) {
    const digit = testNumDigits[i];
    if (!digitCounts.has(digit)) {
      return false;
    }
    digitCounts.set(digit, digitCounts.get(digit) - 1);
    if (digitCounts.get(digit) < 0) {
      return false;
    }
  }

  return true;
}

// always picks arr.length
function getPerms(arr) {
  if (arr.length < 2) {
    return [arr];
  }

  const perms = [];
  for (let i = 0; i < arr.length; i++) {
    const first = arr.splice(i, 1);
    const remainingPerms = getPerms(arr);
    for (let j = 0; j < remainingPerms.length; j++) {
      perms.push(first.concat(remainingPerms[j]));
    }

    arr.splice(i, 0, first[0]);
  }

  return perms;
}

module.exports = { isPermutation, getPerms };

File: python/util/permutation.py
from . import digits


def is_permutation(num, test_num):
  num_digits = digits.get_digits_reversed(num)
  test_num_digits = digits.get_digits_reversed(test_num)

  if len(num_digits) != len(test_num_digits):
    return False

  # create hashmap of digit counts in first array
  digit_counts = {}
  for digit in num_digits:
    if digit_counts.get(digit) is None:
      digit_counts[digit] = 1
    else:
      digit_counts[digit] += 1

  # check digit counts in the second array to make sure they match
  for test_digit in test_num_digits:
    if digit_counts.get(test_digit) is None:
      return False
    else:
      digit_counts[test_digit] -= 1
      if digit_counts[test_digit] < 0:
        return False

  return True


# always picks len(arr)
def get_perms(arr):
  if len(arr) < 2:
    return [arr]

  perms = []
  for i in range(len(arr)):
    first = arr.pop(i)
    remaining_perms = get_perms(arr)
    for j in range(len(remaining_perms)):
      perms.append([first] + remaining_perms[j])

    arr.insert(i, first)

  return perms